File: src/serializer/flat/collect-scopes-in-order.ts
import type { Scope } from "../../ir/model.js";

export function collectScopesInOrder(root: Scope): Scope[] {
  const out: Scope[] = [];
  function visit(s: Scope) {
    out.push(s);
    for (const c of s.childScopes) {
      visit(c);
    }
  }
  visit(root);
  return out;
}

File: crates/unsnarl-emitter-ir/src/serializer/flat/collect_scopes_in_order.rs
//! Pre-order DFS over the scope tree starting at `root`.

use unsnarl_ir::{IrArena, ScopeId};

pub fn collect_scopes_in_order(arena: &IrArena, root: ScopeId) -> Vec<ScopeId> {
    let mut out = Vec::new();
    visit(arena, root, &mut out);
    out
}

fn visit(arena: &IrArena, scope: ScopeId, out: &mut Vec<ScopeId>) {
    out.push(scope);
    for &child in &arena.scopes[scope].child_scopes {
        visit(arena, child, out);
    }
}

#[cfg(test)]
#[path = "collect_scopes_in_order_test.rs"]
mod collect_scopes_in_order_test;
